File: analyzers/tests/SonarAnalyzer.Test/TestCases/ExpectedExceptionAttributeShouldNotBeUsed.NUnit.vb
Imports System
Imports NUnit.Framework

Public Class ExceptionTests

    ' Noncompliant@+2
    <Test>
    <ExpectedException(GetType(DivideByZeroException))>
    Public Sub ExpectedExceptionAttrbutesShouldNotBeUsedOnSub()
        Dim instance As Integer = 42
        instance.ToString()
    End Sub

    ' Noncompliant@+2
    <Test>
    <ExpectedException(GetType(DivideByZeroException))>
    Public Function ExpectedExceptionAttrbutesShouldNotBeUsedOnFunction() As String
        Dim instance As Integer = 42
        Return instance.ToString()
    End Function

    ' Compliant - one line
    <Test>
    <ExpectedException(GetType(DivideByZeroException))>
    Public Sub ExpectedExceptionAttrbuteAllowedForSingleLineSub()
        Throw New NotImplementedException()
    End Sub

    ' Compliant - one line
    <Test>
    <ExpectedException(GetType(DivideByZeroException))>
    Public Function ExpectedExceptionAttrbuteAllowedForSingleLineFunction() As Integer
        Return 42
    End Function

End Class

Public Interface ISomeInterface
    Sub SomeSub()
End Interface

Public MustInherit Class AbstractClass
    Protected MustOverride Function AbstractFunction() As Boolean
End Class

' https://github.com/SonarSource/sonar-dotnet/issues/8300
Class Repro_8300
    <Test>
    <ExpectedException(GetType(InvalidOperationException))>
    Public Sub AssertInFinally()
        Console.ForegroundColor = ConsoleColor.Red
        Try
            Throw New InvalidOperationException()
        Finally
            Assert.AreEqual(ConsoleColor.Black, Console.ForegroundColor)
        End Try
    End Sub
End Class

File: analyzers/tests/SonarAnalyzer.Test/TestCases/ExpectedExceptionAttributeShouldNotBeUsed.NUnit.cs
using System;
using NUnit.Framework;

namespace Tests.Diagnostics
{
    [TestFixture]
    class Program
    {
        [Test]
        [NUnit.Framework.ExpectedException(typeof(ArgumentNullException))]  // Noncompliant
        public void TestFoo2()
        {
            var x = true;
            x.ToString();
        }

        [Test]
        [NUnit.Framework.ExpectedException(typeof(ArgumentNullException))]  // Compliant - one line
        public void TestFoo4()
        {
            new object().ToString();
        }

        [Test]
        [NUnit.Framework.ExpectedException(typeof(ArgumentNullException))]  // Compliant - one line
        public string TestFoo6() => new object().ToString();

        [Test]
        public void TestFoo8()
        {
            bool callFailed = false;
            try
            {
                //...
            }
            catch (ArgumentNullException)
            {
                callFailed = true;
            }
        }

        [Test]
        [NUnit.Framework.ExpectedException(typeof(ArgumentNullException))]
        public void NoBody() // Error [CS0501,CS1002]
    }

    // https://github.com/SonarSource/sonar-dotnet/issues/8300
    class Repro_8300
    {
        [Test]
        [NUnit.Framework.ExpectedException(typeof(InvalidOperationException))] // Compliant - using ExpectedException makes the test more readable
        public void AssertInFinally()
        {
            Console.ForegroundColor = ConsoleColor.Red;
            try
            {
                throw new InvalidOperationException();
            }
            finally
            {
                Assert.AreEqual(ConsoleColor.Black, Console.ForegroundColor);
            }
        }
    }
}
